app/controllers/optIn/ChooseYearController.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.optIn
+
+import auth.{FrontendAuthorisedFunctions, MtdItUser}
+import config.featureswitch.FeatureSwitching
+import config.{AgentItvcErrorHandler, FrontendAppConfig, ItvcErrorHandler}
+import controllers.agent.predicates.ClientConfirmedController
+import forms.optIn.ChooseTaxYearForm
+import models.incomeSourceDetails.TaxYear
+import models.optin.ChooseTaxYearViewModel
+import play.api.Logger
+import play.api.i18n.I18nSupport
+import play.api.mvc._
+import services.optIn.OptInService
+import utils.AuthenticatorPredicate
+import views.html.optIn.ChooseTaxYearView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class ChooseYearController @Inject()(val optInService: OptInService,
+                                     val view: ChooseTaxYearView,
+                                     val authorisedFunctions: FrontendAuthorisedFunctions,
+                                     val auth: AuthenticatorPredicate)
+                                    (implicit val appConfig: FrontendAppConfig,
+                                          mcc: MessagesControllerComponents,
+                                          val ec: ExecutionContext,
+                                          val itvcErrorHandler: ItvcErrorHandler,
+                                          val itvcErrorHandlerAgent: AgentItvcErrorHandler)
+  extends ClientConfirmedController with FeatureSwitching with I18nSupport {
+
+  private val errorHandler = (isAgent: Boolean) => if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+
+  private def withRecover(isAgent: Boolean)(code: => Future[Result])(implicit mtdItUser: MtdItUser[_]): Future[Result] = {
+    code.recover {
+      case ex: Exception =>
+        Logger("application").error(s"request failed :: $ex")
+        errorHandler(isAgent).showInternalServerError()
+    }
+  }
+
+  def show(isAgent: Boolean = false): Action[AnyContent] = auth.authenticatedAction(isAgent) {
+    implicit user =>
+      withRecover(isAgent) {
+        Future.successful(Ok(view(ChooseTaxYearForm(optInService.availableOptInTaxYear().map(_.toString)), viewModel(isAgent))))
+      }
```

Comment:
I can see several calls to `optInService.availableOptInTaxYear()` which will probably necessitate several API calls. I presume this is addressed with the next PR.

Reply:
yes we fetch saved from repository if exists or make api call

Reply:
OK, but if we put repo calls where we currently have API calls, we'll be making several repo calls, so effectively it will have the same issue.

I probably should have said that I meant that the APIs should be refactored so that the data can be shared to where it is required without having to make repeated calls for one page view. I can see that that's been done in the Service PR, so I'll mark this as done.